stan/math/prim/mat/vectorize/apply_scalar_binary.hpp

```diff
@@ -0,0 +1,156 @@
+        return_t result(x.rows(), x.cols());
+        for (int i = 0; i < x.size(); ++i)
+          result(i) = apply_scalar_binary<F, T1, T2>::apply(x(i), y(i));
+        return result;
+      }
+    };
+
+    template <typename F, typename T1, typename T2, int R1, int C1,
+              int R2, int C2>
+    struct apply_scalar_binary<F, Eigen::Block<Eigen::Matrix<T1, R1, C1> >,
+                               Eigen::Block<Eigen::Matrix<T2, R2, C2> > > {
+      /**
+       * Return type, which is calculated recursively as an Eigen::Matrix
+       * of the return type of the contained types T1 and T2.
+       */
+      typedef Eigen::Matrix<typename apply_scalar_binary<F, T1, T2>
+                                ::return_t, R1, C1> return_t;
+
+      /**
+       * Apply the function specified by F elementwise to the
+       * specified argument. This is defined recursively through this
+       * class by passing the arguments as matrices to the
+       * <code>Eigen::Matrix</code> specialization.
+       *
+       * @param x <code>Eigen::Block</code> argument container.
+       * @param y <code>Eigen::Block argument</code> container.
+       * @return Elementwise application of F to the elements of the
+       * containers.
+       */
+      static inline return_t apply(
+          const Eigen::Block<Eigen::Matrix<T1, R1, C1> >& x,
+          const Eigen::Block<Eigen::Matrix<T2, R2, C2> >& y) {
+        return apply_scalar_binary<F, Eigen::Matrix<T1, R1, C1>,
+                                   Eigen::Matrix<T2, R2, C2> >::apply(
+            static_cast<Eigen::Matrix<T1, R1, C1> >(x),
+            static_cast<Eigen::Matrix<T2, R2, C2> >(y));
+      }
+    };
+
+    template <typename F, typename T1, typename T2>
+    struct apply_scalar_binary<F, std::vector<T1>, std::vector<T2> > {
+      /**
+       * Return type, which is calculated recursively as a standard
+       * vector of the return type of the contained types T1 and T2.
+       */
+      typedef typename std::vector<typename apply_scalar_binary<F, T1, T2>
+                                       ::return_t> return_t;
+
+      /**
+       * Apply the function specified by F elementwise to the
+       * specified arguments.  This is defined recursively through this
+       * class applied to elements of types T1 and T2.
+       *
+       * @param x Argument container.
+       * @param y Argument container.
+       * @return Elementwise application of F to the elements of the
+       * containers.
+       */
+      static inline return_t apply(const std::vector<T1>& x,
+                                   const std::vector<T2>& y) {
+        using stan::math::check_size_match;
+
+        check_size_match("binary vectorization", "x", x.size(),
+                         "y", y.size());
+        return_t fx(x.size());
+        for (size_t i = 0; i < x.size(); ++i)
+          fx[i] = apply_scalar_binary<F, T1, T2>::apply(x[i], y[i]);
+        return fx;
+      }
+    };
+  }
```

Comment:
If I'm following correcty, you only deal with vectorization when both arguments are containers of the same type?  The same type and shape and size is OK, but we need the scalar/container and container/scalar versions or this will be very confusing for users.  We can get this request in as is if that's easier.

I'd think they'd jut take implementations like

```
static inline return-t apply(const std::vector<T1>& x, const T2& y) {
  return_t fx(x.size());
  for (size_t i = 0; i < x.size(); ++i)
    fx[i] = apply_scalar_binary<F, T1, T2>::apply(x[i], y);
  return fx;  
}
```

Is there a problem with return type or ambiguities in calls if you try it?  To get rid of the ambiguities is going to require enable-if (to enable `int`, `double`, `var`, and `fvar<U>` or disable-if (to prevent containers).

Reply:
I think we discussed this on Discourse. The problem was that the `mix` tests were failing for the case of `int, std::vector<fvar<var> >` because for `x = 0, y = -4`, taking the second derivative of `std::vector<fvar<var> >` returned `nan` whereas taking the second derivative `fvar<var>` returned `-Inf`. 

Maybe it'll work fine now, but it was easier at the time to drop broadcasting and create this pull request. It's definitely something that should be added.

Reply:
OK, we can wait on the more general case, then.

Reply:
An intermediate pull request (not for here) would be to do `pow()` the right way.  It's the binary operator everyone wants to use and usually the exponent is a scalar.  Really in that case, it'd probably be good to write better derivatives for `pow(vector, real)`.  So the right way to do these is going to start looking like the density functions.